### PR TITLE
Minion

### PR DESCRIPTION
**Pull request by femto**
**Features**
Minion Design

Check examples/smart_minion/brain.py
you can try all different example, I just comment them out.
```
obs, score, *_ = await brain.step(query="what's the solution 234*568")
  # obs, score,*_ = await brain.step(query="what's the solution for  game of 24 for 4 3 9 8")
  # obs, score,*_ = await brain.step(query="what's the solution for  game of 24 for 2 5 11 8")
  # obs, score,*_ = await brain.step(query="solve x=1/(1-beta^2*x) where beta=0.85")

  # obs, score, *_ = await brain.step(
  #     query="Write a 500000 characters novel named 'Reborn in Skyrim'. "
  #           "Fill the empty nodes with your own ideas. Be creative! Use your own words!"
  #           "I will tip you $100,000 if you write a good novel."
  #           "since the novel is very long, you may need to divide into subtasks"
  # )
  #     obs, score, *_ = await brain.step(query=
  #     """
  #     2024阿里巴巴全球数学竞赛
  #
  # 问题1
  #
  # 几位同学假期组成一个小组去某市旅游．该市有6座塔，它们的位置分别为A，B，C，D，B，F。
  #
  # 同学们自由行动一段时间后，每位同学都发现，自己在所在的位置只能看到位于A，B，C，D 处的四座塔，而看不到位于E 和F的塔。已知
  #
  # (1）同学们的位置和塔的位置均视为同一平面上的点，且这些点彼此不重合：
  #
  # (2) A，B，C，D，E，F中任意3点不共线：
  #
  # (3） 看不到塔的唯一可能就是视线被其它的塔所阻挡，例如，如果某位同学所在的位置P 和A，B 共线，且A 在线段PB上，那么该同学就看不到位于B 处的塔。
  #
  # 请问，这个旅游小组最多可能有多少名同学？
  #
  # (A)3 (B) 4 (C)6 (D) 12
  #     """)
  print(obs)
```

Depends on [intercode](https://pypi.org/project/intercode-bench/)
```
git clone https://github.com/princeton-nlp/intercode
docker build -t intercode-python -f docker/python.Dockerfile .
```
make sure container name  intercode-python_ic_ctr  is listenning on 3006, the PythonEnv seemed start the container, but can't map port
```
docker run -d -p 3006:3006 --name intercode-python_ic_ctr intercode-python
```

**Feature Docs**

**Influence**

**Result**

**Other**

### CODE REVIEW

File: examples/smart_minion/brain.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+from metagpt.minion.brain import Brain
+
+
+async def smart_brain():
+    brain = Brain()
+
+    obs, score, *_ = await brain.step(query="what's the solution 234*568")
+    # obs, score,*_ = await brain.step(query="what's the solution for  game of 24 for 4 3 9 8")
+    # obs, score,*_ = await brain.step(query="what's the solution for  game of 24 for 2 5 11 8")
```
Comment (didiforgithub) [inline examples/smart_minion/brain.py:18]: 单纯好奇，这个套方法做Game24正确率高吗

File: metagpt/minion/minion.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+        default="some task type",
+        description="type of this task",
+    )
+    task_params: str = Field(
+        default="{}",
+        description="a json dictionary of task parameters and values",
+    )
+
+
+ASK_PROMPT = """context:
```
Comment (didiforgithub) [inline metagpt/minion/minion.py:74]: 数量较多的 PROMPT 与 TEMPLATE 不然另开一个文件？

File: metagpt/minion/minion.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+{input.short_context}
+instruction:
+{input.instruction}
+query_type:
+{input.query_type}
+query:
+{input.query}
+"""
+
+ASK_PROMPT2 = """context:
```
Comment (didiforgithub) [inline metagpt/minion/minion.py:84]: 能否换一个命名，阐述ASK_PROMPT与ASK_PROMPT2的差异

File: metagpt/minion/minion.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+        filtered_registry = {
+            key: value
+            for key, value in MINION_REGISTRY.items()
+            if key != "smart" and key != "score" and key != "plan" and key != "multi_plan"
+        }
+        filled_template = choose_template.render(minions=filtered_registry, input=self.input)
+
+        plan = await ActionNode.from_pydantic(Plan).fill(context=filled_template, llm=self.brain.llm, schema="raw")
+
+        plan
```
Comment (didiforgithub) [inline metagpt/minion/minion.py:324]: 这里是不是多写了，以及下面大段注释可以删一下

File: metagpt/minion/minion.py [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+                            os.system('python -m pip install sympy')
+                            """,
+                example="",
+            )
+            node = await node.fill(
+                context=(
+                    ASK_PROMPT
+                    + """
+
+also please check previous error, do the modification according to previous error if there's previous error.
```
Comment (didiforgithub) [inline metagpt/minion/minion.py:386]: pre commit 格式化一下

File: metagpt/minion/brain.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+class Brain:
+    def __init__(self, id=None, memory=None, llm=LLM()):
+        self.id = id or uuid.uuid4()
+        self.minds = {}
+        self.add_mind(
+            Mind(
+                id="left_mind",
+                description="""
+I'm the left mind, adept at logical reasoning and analytical thinking. I excel in tasks involving mathematics, language, and detailed analysis. My capabilities include:
```
Comment (didiforgithub) [inline metagpt/minion/brain.py:76]: 格式化

File: examples/smart_minion/brain.py (hunk 1 above)
Comment (femto) [reply examples/smart_minion/brain.py:18]: 你自己试啊，我这些都是可以跑的。
我这两个例子都是100%

**Comment by geekan** [comment]
This work may require further refactoring, or be integrated into aflow and other work. This PR will be closed first. If necessary, we will discuss it separately.

**Pull request closed without merging**